core/src/main/java/jenkins/model/identity/InstanceIdentityProvider.java

```diff
@@ -0,0 +1,226 @@
+ *
+ * Copyright (c) 2016, CloudBees Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package jenkins.model.identity;
+
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import java.security.KeyPair;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.DSAPrivateKey;
+import java.security.interfaces.DSAPublicKey;
+import java.security.interfaces.ECPrivateKey;
+import java.security.interfaces.ECPublicKey;
+import java.security.interfaces.RSAPrivateKey;
+import java.security.interfaces.RSAPublicKey;
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * A source of instance identity.
+ *
+ * @param <PUB>  the type of public key.
+ * @param <PRIV> the type of private key.
+ * @since FIXME
+ */
+public abstract class InstanceIdentityProvider<PUB extends PublicKey, PRIV extends PrivateKey> implements
+        ExtensionPoint {
+    /**
+     * RSA keys.
+     */
+    public static final KeyTypes<RSAPublicKey, RSAPrivateKey> RSA =
+            new KeyTypes<>(RSAPublicKey.class, RSAPrivateKey.class);
+    /**
+     * DSA keys.
+     */
+    public static final KeyTypes<DSAPublicKey, DSAPrivateKey> DSA =
+            new KeyTypes<>(DSAPublicKey.class, DSAPrivateKey.class);
+    /**
+     * EC keys
+     */
+    public static final KeyTypes<ECPublicKey, ECPrivateKey> EC =
+            new KeyTypes<>(ECPublicKey.class, ECPrivateKey.class);
+
+    /**
+     * Gets the {@link KeyPair} that comprises the instance identity.
+     *
+     * @return the {@link KeyPair} that comprises the instance identity. {@code null} could technically be returned in
+     * the event that a keypair could not be generated, for example if the specific key type of this provider
+     * is not permitted at the required length by the JCA policy.
+     */
+    @CheckForNull
+    public abstract KeyPair getKeyPair();
```

Comment:
Might want to make these methods `protected` now, assuming you expect callers to go via `KeyTypes`.